### DISCUSSION TITLE
All modifiers and the words they modify

### DISCUSSION THREAD

**Discussion by webecker25272** [explosion/spaCy#7287] [Help: Coding & Implementations] [answered]
Hi there--
I have a corpus of of super small docs (usually half a sentence each).  I'm trying to figure out the best way to traverse the dependency parser.

My goal is to get all the adjectives in my doc, and retrieve the *chunked* noun which the adjective refers to.  

Example input:
`This home is in a lovely area which is full of interesting shops`

Desired output:
`[{'adj' : 'lovely', 'noun' : 'area'}, {'adj' : 'interesting', 'noun' : 'shops'}]`

Something like that...

What I have so far is something like this:
```
for clause in clauses:
    doc = nlp(clause)
    for token in doc:
        if token.pos_ == 'ADJ':
            for anc in token.ancestors:
                for ch in anc.children:
                    if ch.pos_ == 'NOUN':
                        print([token, ch])
```
where in the last line I would append the matches to the dictionary.. just not sure what structure I want the data in yet, so don't worry about that part.

**QUESTIONS:**
1) is the the best way to traverse the tree?  Am I accessing the adjective's noun properly?  It seems really contrived, but I can't find a better way to do it.
2) I can't access `noun_chunks` once I've drilled in past `for token in doc:`.  Is there any way to retrieve the chunks?

I feel like it can't be this hard... surely I'm overthinking it right?

THANKS!!

**Comment by adrianeboyd** [comment, marked as answer]
I think this would be a good case for the new `DependencyMatcher`: https://spacy.io/usage/rule-based-matching#dependencymatcher

Something like this would be a starting point, which finds all cases where an adjective has a noun as an ancestor:

```python
import spacy
from spacy.matcher import DependencyMatcher

nlp = spacy.load("en_core_web_sm")
matcher = DependencyMatcher(nlp.vocab)

pattern = [
    {
        "RIGHT_ID": "anchor_noun",
        "RIGHT_ATTRS": {"POS": "NOUN"}
    },
    {
        "LEFT_ID": "anchor_noun",
        "REL_OP": ">>",
        "RIGHT_ID": "mod_adj",
        "RIGHT_ATTRS": {"POS": "ADJ"},
    },
]

matcher.add("ADJ_NOUN", [pattern])
doc = nlp("a red, white, and yellow house which is very colorful")
matches = matcher(doc)
for i, (match_id, token_ids) in enumerate(matches):
    print("Match", i, nlp.vocab.strings[match_id])
    for j in range(len(token_ids)):
        print(pattern[j]["RIGHT_ID"] + ":", doc[token_ids[j]].text)

```

Output:

```none
Match 0 ADJ_NOUN
anchor_noun: house
mod_adj: red
Match 1 ADJ_NOUN
anchor_noun: house
mod_adj: white
Match 2 ADJ_NOUN
anchor_noun: house
mod_adj: yellow
Match 3 ADJ_NOUN
anchor_noun: house
mod_adj: colorful
```

The next step would be to use the noun chunks to filter out the matches where the tokens don't all fall in the same noun chunk, like `colorful` above.

**Reply by webecker25272** [reply]
@adrianeboyd  thank you so much!! I will experiment with this.  Very grateful for your response.
